src/server.ts

```diff
@@ -47,128 +39,180 @@ export interface ServerOptions {
-        request.reject(400, 'Unsupported protocol.');
         return;
       }
 
       // accept connection
-      const connection: Connection = request.accept(GRAPHQL_SUBSCRIPTIONS, request.origin);
+      request.resume();
 
       // Regular keep alive messages if keepAlive is set
       if (keepAlive) {
         const keepAliveTimer = setInterval(() => {
-          if (connection && connection.state === 'open') {
-            this.sendKeepAlive(connection);
+          if (request.readyState === WebSocket.OPEN) {
+            this.sendKeepAlive(request);
           } else {
             clearInterval(keepAliveTimer);
           }
         }, keepAlive);
       }
 
       const connectionSubscriptions: ConnectionSubscriptions = {};
-      connection.on('message', this.onMessage(connection, connectionSubscriptions, request));
-      connection.on('close', this.onClose(connection, connectionSubscriptions));
+      request.on('message', this.onMessage(request, connectionSubscriptions));
+      request.on('close', () => {
+        this.onClose(request, connectionSubscriptions)();
+
+        if (this.onDisconnect) {
+          this.onDisconnect(request);
+        }
+      });
     });
   }
 
-  // TODO test that this actually works
-  private onClose(connection: Connection, connectionSubscriptions: ConnectionSubscriptions) {
+  private unsubscribe(connection: WebSocket, handleId: number) {
+    this.subscriptionManager.unsubscribe(handleId);
+
+    if (this.onUnsubscribe) {
+      this.onUnsubscribe(connection);
+    }
+  }
+
+  private onClose(connection: WebSocket, connectionSubscriptions: ConnectionSubscriptions) {
     return () => {
-      Object.keys(connectionSubscriptions).forEach( (subId) => {
-        this.subscriptionManager.unsubscribe(connectionSubscriptions[subId]);
+      Object.keys(connectionSubscriptions).forEach((subId) => {
+        this.unsubscribe(connection, connectionSubscriptions[subId]);
         delete connectionSubscriptions[subId];
       });
-    }
+    };
   }
 
-  private onMessage(connection: Connection, connectionSubscriptions: ConnectionSubscriptions, webSocketRequest: WebSocketRequest) {
-    return  (message: IMessage) => {
+  private onMessage(connection: WebSocket, connectionSubscriptions: ConnectionSubscriptions) {
+    return (message: any) => {
       let parsedMessage: SubscribeMessage;
       try {
-        parsedMessage = JSON.parse(message.utf8Data);
+        parsedMessage = JSON.parse(message);
       } catch (e) {
-        this.sendSubscriptionFail(connection, null, { errors: [{ message: e.message }] });
+        this.sendSubscriptionFail(connection, null, {errors: [{message: e.message}]});
         return;
       }
 
       const subId = parsedMessage.id;
       switch (parsedMessage.type) {
+        case INIT:
+          let onConnectPromise = Promise.resolve(true);
+          if (this.onConnect) {
+            onConnectPromise = Promise.resolve(this.onConnect(parsedMessage.payload, connection));
```

Comment:
Would it be simpler to get rid of initResult and just resolve the initPromise with initResult directly? You don't use initResult anywhere that's not already inside a initPromise.then...